### PR TITLE
generalized chat sft prompt

### PR DESCRIPTION
**Pull request by yidong72**
# What does this PR do ?

In this PR, it genialized the chat SFT dataset that it can use customized turn start/end tokens by using chat_prompt_tokens config. e.g.
```yaml
    chat_prompt_tokens:  # special tokens for the chat prompts, a dictionary of {token_type: token}. note that some tokenizer may combine the characters at the junction between {end_of_turn}{turn_start}. e.g. '<im end><im start>', the '><' sometimes is merged to be a single token. This is not supported, try to avoid
      system_turn_start: '<extra_id_0>'
      turn_start: '<extra_id_1>'
      label_start: '<extra_id_2>'
      end_of_turn: "\x0A"  # \0x0A is '\n'
      end_of_name: "\x0A"  # \0x0A is '\n'
```
after this change, the LM is not required to have "extra_id" special tokens any more to use chat SFT dataset. In this PR, also expanded the unit test to cover more LM tokenizers.

Another feature added is to overwrite the prompt_template config with the chat prompt format.

### CODE REVIEW

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 1]
```diff
@@ -228,45 +282,39 @@ def preprocess(source: dict, tokenizer: TokenizerSpec, extra_id_2_token_id: int,
...
-        ids_2 = self.tokenizer.text_to_ids('<extra_id_1>')
+        LABEL_START = self.special_tokens['label_start']
+        id1 = self.tokenizer.text_to_ids(PREFIX_STR)
+        id2 = self.tokenizer.text_to_ids(PREFIX_STR + LABEL_START)
+        self.label_start_tokens = id2[len(id1) :]
+        ids_1 = self.tokenizer.text_to_ids(PREFIX_STR + '\n')
+        ids_2 = self.tokenizer.text_to_ids(PREFIX_STR)
         self.new_line_token_id = ids_1[len(ids_2) :][0]
 
+        ids_1 = self.tokenizer.text_to_ids(PREFIX_STR + self.special_tokens['turn_start'])
```
Comment (Zhilin123) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:328]: Do you need to define id1/2 and ids_1/2 multiple times?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 2]
```diff
@@ -16,26 +16,41 @@
...
+END_NAME_SIGNAL = "\n"  # token to indicate the end of the name. The name can be system, user, assistant, etc.
+SYSTEM_TOKEN = "System" + END_NAME_SIGNAL
 
 TYPE_INSTRUCTION = {
     'TEXT_TO_VALUE': "",
     'VALUE_TO_TEXT': '',
 }
 
 
+def find_small_tensor(small, large):
```
Comment (Zhilin123) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:39]: Maybe a better name is `identify_start_index_of_subsequence(subsequence, sequence)` 
Otherwise, it took me some time to figure out what small and large means here

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml [hunk 3]
```diff
@@ -71,6 +71,11 @@ model:
 
   data:
     chat: False # whether use chatbot data or not
+    chat_prompt_tokens:  # special tokens for the chat prompts, a dictionary of {token_type: token}. note that some tokenizer may combine the characters at the junction between {end_of_turn}{turn_start}. e.g. '<im end><im start>', the '><' sometimes is merged to be a single token. This is not supported, try to avoid
```
Comment (Zhilin123) [inline examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:74]: Maybe a good thing to mention is what these should be for llama2? Otherwise, people might not know what to set it to

**Review by Zhilin123 — APPROVED** [review]
LGTM, minor code style issues

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 2 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py]: changed

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml (hunk 3 above)
Comment (yidong72) [reply examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:74]: llama2 can just use the default. it is user's choice to do whatever here.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 1 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py]: all use the same variable names

File: examples/nlp/language_modeling/megatron_gpt_eval.py [hunk 4]
```diff
@@ -167,7 +168,11 @@ def remove_padded_prompts(response, nb_paddings):
 def main(cfg) -> None:
 
     # trainer required for restoring model parallel models
-    trainer = Trainer(strategy=NLPDDPStrategy(), **cfg.trainer, callbacks=[CustomProgressBar()])
+    trainer = Trainer(
+        strategy=NLPDDPStrategy(timeout=datetime.timedelta(seconds=18000)),
```
Comment (yidong72) [inline examples/nlp/language_modeling/megatron_gpt_eval.py:172]: this change is done to make sure the inference serve not time out (NCCL time out error) because of no activities.

**Review by Zhilin123 — APPROVED** [review]
LGTM

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 5]
```diff
@@ -93,6 +99,16 @@ def __init__(
...
+        if special_tokens is None:
+            self.special_tokens = {
+                "system_turn_start": "<extra_id_0>",
+                "turn_start": "<extra_id_1>",
+                "label_start": "<extra_id_2>",
+                "end_of_turn": "\n",
+                "end_of_name": "\n",
+            }
+        else:
+            self.special_tokens = special_tokens
```
Comment (aklife97) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:112]: can we do a check to see if the tokens in special_tokens are tokenizer's special tokens or not? If not (the case with llama), can we just throw a warning that we'll use text as turn tokens which _might_ cause incorrect merging

File: scripts/nlp_language_modeling/sft/preprocessing.py [hunk 6]
```diff
@@ -80,11 +80,12 @@ def parse_conversations(tree_obj):
         raise ValueError(f'unknown role {prompt_obj["role"]}')
     turn = {'value': prompt_obj['text'], 'from': role}
     if 'labels' in prompt_obj:
-        turn['human_labels'] = prompt_obj['labels']
-        for key in turn['human_labels']:
-            value_set = label_values.get(key, set())
-            value_set.add(turn['human_labels'][key]['value'])
-            label_values[key] = value_set
+        # remove human labels
+        # turn['human_labels'] = prompt_obj['labels']
```
Comment (aklife97) [inline scripts/nlp_language_modeling/sft/preprocessing.py:84]: why remove this? wouldnt the naming "label" get overridden and confusing after reward model annotation?

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py [hunk 7]
```diff
@@ -71,6 +74,13 @@ def _modify_config(gpt_cfg, cfg, add_cfg_to_tree=False):
         gpt_cfg.pipeline_model_parallel_size = cfg.model.get('pipeline_model_parallel_size', 1)
         gpt_cfg.pipeline_model_parallel_split_rank = cfg.model.get('pipeline_model_parallel_split_rank', 0)
 
+        if cfg.model.data.get('chat', False):
+            # chat model, overwrite the prompt template
+            prompt_template = get_prompt_template_example(cfg.model.data.chat_prompt_tokens)
```
Comment (aklife97) [inline examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:79]: TODO for later: think what to call this. Dont want people to confuse with the strict prompt_template we use elsewhere

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 8]
```diff
@@ -16,26 +16,76 @@
...
+__all__ = ['GPTSFTChatDataset', 'get_prompt_template_example']
 
-IGNORE_INDEX = -100
-END_SIGNAL = "\n"
-END_NAME_SIGNAL = "\n"
 
-SYSTEM_TOKEN = "<extra_id_0>System\n"
-TURN_TOKEN = "<extra_id_1>"
+PREFIX_STR = (
+    "\x00"  # the prefix string used in the tokenizer to deal with the added empty token for some of the tokenizers
```
Comment (aklife97) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:27]: outside this PR, but I think we should also make this configurable
not all tokenizers use byte-level-fallback

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 9]
```diff
@@ -60,50 +112,69 @@ def _mask_targets(
...
             if not torch.equal(target[cur_idx + 1 : cur_idx + tokenized_len], s_id[1:]):
                 logging.warning("a sentence mismatches the corresponding piece " "in the conversation")
         if i == 0 and (gtype == 'VALUE_TO_TEXT' or gtype is None):
-            # mask the first turn completely to provide at least one turn as context
+            # mask the first turn completely to provide at least one turn as context for the rest
             target[cur_idx : cur_idx + tokenized_len] = IGNORE_INDEX
         elif speaker == mask_role and i == 1 and gtype == 'TEXT_TO_VALUE':
-            # leave the first human tag unmasked
-            target[cur_idx + 1 : cur_idx + tokenized_len] = IGNORE_INDEX
+            # leave the first turn start tag unmasked, servers serves as the end of turn signal
```
Comment (aklife97) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:168]: nit: can we fix the "servers serves" typo? I think it should just be "serves"

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 10]
```diff
@@ -60,50 +112,69 @@ def _mask_targets(
...
+            target[cur_idx + num_turn_start_tokens : cur_idx + tokenized_len] = IGNORE_INDEX
         elif speaker == mask_role and (i > 1):
-            # leave the first human tag unmasked
-            target[cur_idx + 1 : cur_idx + tokenized_len] = IGNORE_INDEX
+            # leave the first turn start tag unmasked, which servers as the end of turn signal
```
Comment (aklife97) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:171]: typo here as well

File: scripts/nlp_language_modeling/sft/data_clean.py [hunk 11]
```diff
@@ -41,7 +41,7 @@ def data_clean(
     )
     if library == 'huggingface':
         tokenizer.add_special_tokens({'additional_special_tokens': ['<extra_id_0>', '<extra_id_1>', '<extra_id_2>']})
-    d = GPTSFTChatDataset(dataset_file, tokenizer, seq_len, 1)
+    d = GPTSFTChatDataset(dataset_file, tokenizer, seq_len, 1, hf_dataset=True)
```
Comment (aklife97) [inline scripts/nlp_language_modeling/sft/data_clean.py:44]: can we change the default back to False?
I dont think we should force hf_dataset, people can set it to True if they want to use it

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 8 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:27]: agree

File: scripts/nlp_language_modeling/sft/preprocessing.py (hunk 6 above)
Comment (aklife97) [reply scripts/nlp_language_modeling/sft/preprocessing.py:84]: so just looking at the data, we cant be sure if its been annotated by reward model or not? can we add a new field for it?
currently I used to match label with human_label to understand that

File: scripts/nlp_language_modeling/sft/data_clean.py (hunk 11 above)
Comment (yidong72) [reply scripts/nlp_language_modeling/sft/data_clean.py:44]: the problem is hf_dataset False will fail because some bug in the `jsontextmap` dataset.  Set this to True is safer.

Comment (aklife97) [reply scripts/nlp_language_modeling/sft/data_clean.py:44]: why was it not there earlier?
might be because of the `mp.set_start_method("spawn", force=True)` ?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 9 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:168]: fixed

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py (hunk 10 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:171]: fixed

File: scripts/nlp_language_modeling/sft/data_clean.py (hunk 11 above)
Comment (yidong72) [reply scripts/nlp_language_modeling/sft/data_clean.py:44]: don't know. but our old dataset is not reliable that's why I implemented the hf dataset.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py [hunk 12]
```diff
@@ -112,16 +183,21 @@ def cannonical_form_formater(cannoical_form):
     return f'<extra_id_2>{cannoical_form}\n'
```
[the remark this answers is not in the document]
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_chat_dataset.py:183]: removed

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 5 above)
Comment (aklife97) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:112]: that is different, the token merge can still happen during multi-turn

what I mean is that if the turn tokens are not special tokens, we just say that there might be an error possible

File: scripts/nlp_language_modeling/sft/preprocessing.py (hunk 6 above)
Comment (yidong72) [reply scripts/nlp_language_modeling/sft/preprocessing.py:84]:
```python
                    if 'label' in item['conversations'][turn]:
                        item['conversations'][turn]['gt_label'] = item['conversations'][turn]['label']
```
In the annotation, I set the human label to `gt_label`

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 5 above)
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:112]: The `header_len` stops at the "end_of_turn". The next token is "turn_start". If the merge happens this assert will catch it. The multiple turn has the same thing. each turn ends with "end_of_turn" and the next token is "turn_start". So this one is enough to catch it. 

Also I don't see the point of just giving a warning which doesn't help the user at all.

File: scripts/nlp_language_modeling/sft/data_clean.py (hunk 11 above)
Comment (yidong72) [reply scripts/nlp_language_modeling/sft/data_clean.py:44]: This is `data_clean.py` script. It won't affect the SFT `GPTSFTChatDataset` initialization which is default to be false. I don't see any risk here.

Comment (aklife97) [reply scripts/nlp_language_modeling/sft/data_clean.py:44]: ah, didnt see this was in a script 

looks good

**Review by aklife97 — APPROVED** [review]
LGTM, thanks!
